File: uci.py
#uci.py 

import chess
from search import pick_move_timed
from tt import tt
from helpers import history_heuristic
import sys

# Force unbuffered output for PyInstaller compatibility
sys.stdout.reconfigure(line_buffering=True) if hasattr(sys.stdout, 'reconfigure') else None
sys.stderr.reconfigure(line_buffering=True) if hasattr(sys.stderr, 'reconfigure') else None

board = chess.Board()
stochastic_mode = False

def uci_print(message):
    """Print with guaranteed flush for UCI communication"""
    print(message, flush=True)

def main():
    global stochastic_mode
    
    while True:
        try:
            line = input()
        except EOFError:
            break

        if line == "uci":
            uci_print("id name Iteration1")
            uci_print("id author Moodlethenoodle")
            uci_print("option name Stochastic type check default false")
            uci_print("option name Hash type spin default 256 min 1 max 2048")
            uci_print("uciok")

        elif line == "isready":
            uci_print("readyok")

        elif line.startswith("setoption"):
            if "name Stochastic" in line:
                if "value true" in line:
                    stochastic_mode = True
                    uci_print("# Stochastic mode enabled")
                else:
                    stochastic_mode = False
                    uci_print("# Stochastic mode disabled")
            elif "name Hash" in line:
                try:
                    parts = line.split()
                    value_idx = parts.index("value") + 1
                    hash_size = int(parts[value_idx])
                    tt.resize(hash_size)
                    uci_print(f"# Hash table resized to {hash_size}MB")
                except (ValueError, IndexError):
                    pass

        elif line == "ucinewgame":
            # Clear transposition table and history on new game
            tt.clear()
            history_heuristic.clear()
            uci_print("# New game - cleared TT and history")

        elif line.startswith("position"):
            if "startpos" in line:
                board.reset()
                if "moves" in line:
                    moves = line.split("moves")[1].strip().split()
                    for move in moves:
                        board.push_uci(move)

            elif "fen" in line:
                fen = line[8:].strip()
                if "moves" in fen:
                    fen, moves_str = fen.split("moves")
                    board.set_fen(fen.strip())
                    for move in moves_str.strip().split():
                        board.push_uci(move)
                else:
                    board.set_fen(fen)

        elif line.startswith("go"):
            parts = line.split()
            movetime = None
            wtime = None
            btime = None
            winc = None
            binc = None
            movestogo = None
            max_depth = None
            
            i = 1
            while i < len(parts):
                if parts[i] == "movetime":
                    movetime = int(parts[i + 1])
                    i += 2
                elif parts[i] == "wtime":
                    wtime = int(parts[i + 1])
                    i += 2
                elif parts[i] == "btime":
                    btime = int(parts[i + 1])
                    i += 2
                elif parts[i] == "winc":
                    winc = int(parts[i + 1])
                    i += 2
                elif parts[i] == "binc":
                    binc = int(parts[i + 1])
                    i += 2
                elif parts[i] == "movestogo":
                    movestogo = int(parts[i + 1])
                    i += 2
                elif parts[i] == "depth":
                    max_depth = int(parts[i + 1])
                    i += 2
                elif parts[i] == "infinite":
                    max_depth = 99
                    i += 1
                else:
                    i += 1
            
            # Calculate time allocation with safety buffer
            time_limit = None
            SAFETY_BUFFER_MS = 1500  # 1.5 second safety buffer
            
            if movetime:
                time_limit = max(movetime / 1000.0 - 0.05, 0.01)  # Just 50ms for movetime
            elif board.turn == chess.WHITE and wtime:
                increment = winc if winc else 0
                moves_left = movestogo if movestogo else 30
                
                # Reserve 1.5 seconds as safety buffer
                safe_time = max(wtime - SAFETY_BUFFER_MS, 100)
                
                # Use 85% of safe time per move + 90% of increment
                time_limit = max((safe_time / 1000.0 / moves_left * 0.85) + (increment / 1000.0 * 0.9), 0.01)
                
            elif board.turn == chess.BLACK and btime:
                increment = binc if binc else 0
                moves_left = movestogo if movestogo else 30
                
                # Reserve 1.5 seconds as safety buffer
                safe_time = max(btime - SAFETY_BUFFER_MS, 100)
                
                # Use 85% of safe time per move + 90% of increment
                time_limit = max((safe_time / 1000.0 / moves_left * 0.85) + (increment / 1000.0 * 0.9), 0.01)
            
            if max_depth is None:
                max_depth = 64  # Search deeper by default
            
            best_move, all_moves = pick_move_timed(board, max_depth, time_limit, stochastic_mode)
            
            if best_move:
                uci_print(f"bestmove {best_move.uci()}")
            else:
                # Fallback to any legal move
                legal_moves = list(board.legal_moves)
                if legal_moves:
                    uci_print(f"bestmove {legal_moves[0].uci()}")

        elif line == "quit":
            break

if __name__ == '__main__':
    main()

File: src/uci.rs
use chess::{Board, ChessMove, Color, MoveGen};
use std::io::{self, BufRead, Write};
use std::str::FromStr;

use crate::search::{pick_move_timed, SearchState};

pub fn run() {
    let stdin = io::stdin();
    let mut stdout = io::stdout();
    
    let mut board = Board::default();
    let mut state = SearchState::new();

    for line in stdin.lock().lines() {
        let line = line.unwrap();
        let parts: Vec<&str> = line.split_whitespace().collect();

        if parts.is_empty() {
            continue;
        }

        match parts[0] {
            "uci" => {
                println!("id name Iteration1");
                println!("id author Moodlethenoodle");
                println!("option name Hash type spin default 256 min 1 max 2048");
                println!("uciok");
                stdout.flush().unwrap();
            }
            "isready" => {
                println!("readyok");
                stdout.flush().unwrap();
            }
            "setoption" => {
                if line.contains("name Hash") {
                    if let Some(value_pos) = parts.iter().position(|&p| p == "value") {
                        if let Some(size_str) = parts.get(value_pos + 1) {
                            if let Ok(size) = size_str.parse::<usize>() {
                                state.tt = crate::tt::TranspositionTable::new(size);
                                println!("# Hash table resized to {}MB", size);
                            }
                        }
                    }
                }
                stdout.flush().unwrap();
            }
            "ucinewgame" => {
                state.tt.clear();
                state.history.clear();
                state.position_history.clear();  // Clear position history for new game
                println!("# New game - cleared TT, history, and position history");
                stdout.flush().unwrap();
            }
            "position" => {
                state.position_history.clear();
                state.halfmove_clock = 0;  // NEW: Reset halfmove clock

                if parts[1] == "startpos" {
                    board = Board::default();
                    
                    if let Some(moves_pos) = parts.iter().position(|&p| p == "moves") {
                        for move_str in &parts[moves_pos + 1..] {
                            state.position_history.push(board.get_hash());
                            
                            if let Ok(mv) = ChessMove::from_san(&board, move_str) {
                                // NEW: Update halfmove clock
                                if state.is_irreversible(&board, mv) {
                                    state.halfmove_clock = 0;
                                } else {
                                    state.halfmove_clock += 1;
                                }
                                board = board.make_move_new(mv);
                            } else if let Ok(mv) = ChessMove::from_str(move_str) {
                                // NEW: Update halfmove clock
                                if state.is_irreversible(&board, mv) {
                                    state.halfmove_clock = 0;
                                } else {
                                    state.halfmove_clock += 1;
                                }
                                board = board.make_move_new(mv);
                            }
                        }
                    }
                } else if parts[1] == "fen" {
                    let moves_pos = parts.iter().position(|&p| p == "moves");
                    let fen_end = moves_pos.unwrap_or(parts.len());
                    let fen_str = parts[2..fen_end].join(" ");
                    
                    if let Ok(new_board) = Board::from_str(&fen_str) {
                        board = new_board;
                        
                        // NEW: Parse halfmove clock from FEN if present
                        let fen_parts: Vec<&str> = fen_str.split_whitespace().collect();
                        if fen_parts.len() >= 5 {
                            if let Ok(halfmove) = fen_parts[4].parse::<u32>() {
                                state.halfmove_clock = halfmove;
                            }
                        }
                        
                        if let Some(moves_pos) = moves_pos {
                            for move_str in &parts[moves_pos + 1..] {
                                state.position_history.push(board.get_hash());
                                
                                if let Ok(mv) = ChessMove::from_san(&board, move_str) {
                                    // NEW: Update halfmove clock
                                    if state.is_irreversible(&board, mv) {
                                        state.halfmove_clock = 0;
                                    } else {
                                        state.halfmove_clock += 1;
                                    }
                                    board = board.make_move_new(mv);
                                } else if let Ok(mv) = ChessMove::from_str(move_str) {
                                    // NEW: Update halfmove clock
                                    if state.is_irreversible(&board, mv) {
                                        state.halfmove_clock = 0;
                                    } else {
                                        state.halfmove_clock += 1;
                                    }
                                    board = board.make_move_new(mv);
                                }
                            }
                        }
                    }
                }
            }
            "go" => {
                let mut movetime = None;
                let mut wtime = None;
                let mut btime = None;
                let mut winc = None;
                let mut binc = None;
                let mut movestogo = None;
                let mut max_depth = None;

                let mut i = 1;
                while i < parts.len() {
                    match parts[i] {
                        "movetime" => {
                            if let Some(val) = parts.get(i + 1) {
                                movetime = val.parse::<u64>().ok();
                            }
                            i += 2;
                        }
                        "wtime" => {
                            if let Some(val) = parts.get(i + 1) {
                                wtime = val.parse::<u64>().ok();
                            }
                            i += 2;
                        }
                        "btime" => {
                            if let Some(val) = parts.get(i + 1) {
                                btime = val.parse::<u64>().ok();
                            }
                            i += 2;
                        }
                        "winc" => {
                            if let Some(val) = parts.get(i + 1) {
                                winc = val.parse::<u64>().ok();
                            }
                            i += 2;
                        }
                        "binc" => {
                            if let Some(val) = parts.get(i + 1) {
                                binc = val.parse::<u64>().ok();
                            }
                            i += 2;
                        }
                        "movestogo" => {
                            if let Some(val) = parts.get(i + 1) {
                                movestogo = val.parse::<u32>().ok();
                            }
                            i += 2;
                        }
                        "depth" => {
                            if let Some(val) = parts.get(i + 1) {
                                max_depth = val.parse::<i32>().ok();
                            }
                            i += 2;
                        }
                        "infinite" => {
                            max_depth = Some(99);
                            i += 1;
                        }
                        _ => i += 1,
                    }
                }

                let time_limit = if let Some(mt) = movetime {
                    Some((mt as f64 / 1000.0 - 0.05).max(0.01))
                } else if board.side_to_move() == Color::White {
                    if let Some(wt) = wtime {
                        let increment = winc.unwrap_or(0) as f64;
                        let moves_left = movestogo.unwrap_or(30) as f64;
                        let safe_time = (wt as i64 - 1500).max(100) as f64;
                        Some(((safe_time / 1000.0 / moves_left * 0.85) + (increment / 1000.0 * 0.9)).max(0.01))
                    } else {
                        None
                    }
                } else {
                    if let Some(bt) = btime {
                        let increment = binc.unwrap_or(0) as f64;
                        let moves_left = movestogo.unwrap_or(30) as f64;
                        let safe_time = (bt as i64 - 1500).max(100) as f64;
                        Some(((safe_time / 1000.0 / moves_left * 0.85) + (increment / 1000.0 * 0.9)).max(0.01))
                    } else {
                        None
                    }
                };

                let depth = max_depth.unwrap_or(64);

                let (best_move, _) = pick_move_timed(&board, depth, time_limit, &mut state);

                if let Some(mv) = best_move {
                    println!("bestmove {}", mv);
                } else {
                    let legal_moves: Vec<ChessMove> = MoveGen::new_legal(&board).collect();
                    if !legal_moves.is_empty() {
                        println!("bestmove {}", legal_moves[0]);
                    }
                }
                stdout.flush().unwrap();
            }
            "quit" => break,
            _ => {}
        }
    }
}